hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/autoassign/AbstractAutoAssignExecutor.java

```diff
@@ -0,0 +1,168 @@
+import org.eclipse.hawkbit.repository.model.TargetFilterQuery;
+import org.eclipse.hawkbit.tenancy.TenantAware;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.PageRequest;
+import org.springframework.data.domain.Pageable;
+import org.springframework.transaction.PlatformTransactionManager;
+import org.springframework.transaction.annotation.Isolation;
+import org.springframework.util.StringUtils;
+
+/**
+ * Abstract implementation of an AutoAssignExecutor
+ */
+public abstract class AbstractAutoAssignExecutor implements AutoAssignExecutor {
+
+    /**
+     * The message which is added to the action status when a distribution set is
+     * assigned to an target. First %s is the name of the target filter.
+     */
+    private static final String ACTION_MESSAGE = "Auto assignment by target filter: %s";
+    
+    /**
+     * Maximum for target filter queries with auto assign DS Maximum for targets
+     * that are fetched in one turn
+     */
+    private static final int PAGE_SIZE = 1000;
+
+    private final TargetFilterQueryManagement targetFilterQueryManagement;
+
+    private final DeploymentManagement deploymentManagement;
+
+    private final PlatformTransactionManager transactionManager;
+
+    private final TenantAware tenantAware;
+
+    /**
+     * Constructor
+     * 
+     * @param targetFilterQueryManagement
+     *            to get all target filter queries
+     * @param deploymentManagement
+     *            to assign distribution sets to targets
+     * @param transactionManager
+     *            to run transactions
+     * @param tenantAware
+     *            to handle the tenant context
+     */
+    protected AbstractAutoAssignExecutor(final TargetFilterQueryManagement targetFilterQueryManagement,
+            final DeploymentManagement deploymentManagement, final PlatformTransactionManager transactionManager,
+            final TenantAware tenantAware) {
+        this.targetFilterQueryManagement = targetFilterQueryManagement;
+        this.deploymentManagement = deploymentManagement;
+        this.transactionManager = transactionManager;
+        this.tenantAware = tenantAware;
+    }
+
+    protected TargetFilterQueryManagement getTargetFilterQueryManagement() {
+        return targetFilterQueryManagement;
+    }
+
+    protected DeploymentManagement getDeploymentManagement() {
+        return deploymentManagement;
+    }
+
+    protected PlatformTransactionManager getTransactionManager() {
+        return transactionManager;
+    }
+
+    protected TenantAware getTenantAware() {
+        return tenantAware;
+    }
+
+    protected void forEachFilterWithAutoAssignDS(final Consumer<TargetFilterQuery> consumer) {
+        Page<TargetFilterQuery> filterQueries;
+        Pageable query = PageRequest.of(0, PAGE_SIZE);
+
+        do {
+            filterQueries = targetFilterQueryManagement.findWithAutoAssignDS(query);
+
+            filterQueries.forEach(consumer);
```

Comment:
great! I would however catch all Runtime exceptions (similar to `JpaSystemManagement#forEachTenant`) in forEach block to continue processing other filter queries in case of a faulty one